### PR TITLE
chore(*): change installation of spacy weights to runtime

### PR DESCRIPTION
**Pull request by KrishPatel13**
**What kind of change does this PR introduce?**

feature

**Summary**
Motivation: https://github.com/OpenAdaptAI/OpenAdapt/issues/461

**Checklist**
* [x] My code follows the style guidelines of OpenAdapt
* [x] I have performed a self-review of my code
* [x] If applicable, I have added tests to prove my fix is functional/effective
* [x] I have linted my code locally prior to submission
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] I have made corresponding changes to the documentation (e.g. README.md, requirements.txt)
* [x] New and existing unit tests pass locally with my changes

**How can your code be run and tested?**
```
pip uninstall `en_core_web_trf`
# Follow Setup Instructions in README.md
pytest -s
```

**Other information**

### CODE REVIEW

**Comment by KrishPatel13** [comment]
This is how pytest will look like when we do not have the required spacy weight model installed:

```
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> pytest
============================================================================== test session starts ==============================================================================
platform win32 -- Python 3.10.11, pytest-7.1.3, pluggy-1.2.0
rootdir: P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt
plugins: anyio-3.7.1, Faker-19.2.0, cov-2.10.0
collected 10 items / 1 error

==================================================================================== ERRORS =====================================================================================
________________________________________________________________ ERROR collecting tests/openadapt/test_scrub.py _________________________________________________________________
tests\openadapt\test_scrub.py:9: in <module>
    from openadapt import config, scrub
openadapt\scrub.py:25: in <module>
    NLP_ENGINE_TRF = SCRUB_PROVIDER_TRF.create_engine()
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\presidio_analyzer\nlp_engine\nlp_engine_provider.py:91: in create_engine
    engine = nlp_engine_class(nlp_engine_opts)
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\presidio_analyzer\nlp_engine\spacy_nlp_engine.py:36: in __init__        
    self.nlp = {
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\presidio_analyzer\nlp_engine\spacy_nlp_engine.py:37: in <dictcomp>      
    lang_code: spacy.load(model_name, disable=["parser"])
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\spacy\__init__.py:51: in load
    return util.load_model(
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\spacy\util.py:472: in load_model
    raise IOError(Errors.E050.format(name=name))
E   OSError: [E050] Can't find model 'en_core_web_trf'. It doesn't seem to be a Python package or a valid path to a data directory.
=============================================================================== warnings summary ================================================================================ 
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pycountry\__init__.py:10
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pycountry\__init__.py:10: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('mpl_toolkits')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================ short test summary info ============================================================================ 
ERROR tests/openadapt/test_scrub.py - OSError: [E050] Can't find model 'en_core_web_trf'. It doesn't seem to be a Python package or a valid path to a data directory.
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! Interrupted: 1 error during collection !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! 
========================================================================= 3 warnings, 1 error in 6.29s ========================================================================== 
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> 
```

**Comment by KrishPatel13** [comment]
@abrichr Ready for review!

**Comment by KrishPatel13** [comment]
Some Screenshots:

image

**Comment by KrishPatel13** [comment]
This is be downloaded in runtime. So whenever the user had to use the scrub code  either via record or pytest or visualization, then it will check for the installation for the spacy weight and if not found it will install and then it will continue.

**Review by Mustaballer — APPROVED** [review]
lgtm!

**Comment by abrichr** [comment]
Thanks @KrishPatel13 !

What do you think about the approach mentioned here https://github.com/explosion/spaCy/issues/4592#issuecomment-704373657 :

```
import spacy

spacy_model_name = 'de_core_news_sm'
if not spacy.util.is_package(spacy_model_name):
    spacy.cli.download(spacy_model_name)
```

Also, what do you think about skipping tests which depend on spacy if the model is not installed with `@pytest.mark.skipif` ?

**Comment by KrishPatel13** [comment]
Good point, I will include that.

**Comment by KrishPatel13** [comment]
We will ahve to skip all of the test functions in test_scrub if the required spacy is not installed. Hence,

A better approach: https://docs.pytest.org/en/7.1.x/how-to/skipping.html#:~:text=Skip%20all%20test%20functions%20of%20a%20class%20or%20module&text=If%20you%20want%20to%20skip,skipif(...)

image

**Comment by KrishPatel13** [comment]
```
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> pytest
================================ test session starts ================================
platform win32 -- Python 3.10.11, pytest-7.1.3, pluggy-1.2.0
rootdir: P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt
plugins: anyio-3.7.1, Faker-19.2.0, cov-2.10.0
collected 25 items

tests\openadapt\test_crop.py .                                                 [  4%]
tests\openadapt\test_events.py .......                                         [ 32%]
tests\openadapt\test_scrub.py sssssssssssssss                                  [ 92%]
tests\openadapt\test_summary.py ..                                             [100%]

================================= warnings summary ==================================
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pycountry\__init__.py:10
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pycountry\__init__.py:10: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    import pkg_resources

C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('mpl_toolkits')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
=============== 10 passed, 15 skipped, 3 warnings in 96.82s (0:01:36) =============== 
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> 
```

**Comment by KrishPatel13** [comment]
```
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> pip uninstall en_core_web_trf
WARNING: Skipping en_core_web_trf as it is not installed.
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> pytest
================================ test session starts ================================
platform win32 -- Python 3.10.11, pytest-7.1.3, pluggy-1.2.0
rootdir: P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt
plugins: anyio-3.7.1, Faker-19.2.0, cov-2.10.0
collected 25 items

tests\openadapt\test_crop.py .                                                 [  4%]
tests\openadapt\test_events.py .......                                         [ 32%]
tests\openadapt\test_scrub.py sssssssssssssss                                  [ 92%]
tests\openadapt\test_summary.py ..                                             [100%]

================================= warnings summary ==================================
C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pycountry\__init__.py:10
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pycountry\__init__.py:10: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html      
    import pkg_resources

C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('mpl_toolkits')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871
  C:\Users\Krish Patel\AppData\Local\pypoetry\Cache\virtualenvs\openadapt-DSRh12US-py3.10\lib\site-packages\pkg_resources\__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('sphinxcontrib')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
==================== 10 passed, 15 skipped, 3 warnings in 8.55s ===================== 
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> pip show en_core_web_trf
WARNING: Package(s) not found: en_core_web_trf
(openadapt-py3.10) PS P:\OpenAdapt AI - MLDS AI\cloned_repo\test_other\OpenAdapt> imestamp=
```

**Comment by KrishPatel13** [comment]
@abrichr Ready for merging ;-)

addressed your comment

File: tests/openadapt/test_scrub.py [hunk 1]
```diff
@@ -2,11 +2,17 @@
...
 
 from PIL import Image
+import pytest
+import spacy
 
-from openadapt import config, scrub
+from openadapt import config
+
+if not spacy.util.is_package(config.SPACY_MODEL_NAME):
+    pytestmark = pytest.mark.skip(reason="SpaCy model not installed.")
```
Comment (abrichr) [inline tests/openadapt/test_scrub.py:13]: This does not appear to be used anywhere, is that intentional?

Comment (KrishPatel13) [reply tests/openadapt/test_scrub.py:13]: It is used when running pytest.

Basically, when the user does not have the Spacy Model installed then all of the tests in the file will be skipped.

Comment (KrishPatel13) [reply tests/openadapt/test_scrub.py:13]: If the model is installed then, it will import scrub and continue to run the tests

**Pull request merged**